child_process: support numeric signal argument to kill.

gerrard00 opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
child_process

##### Description of change
<!-- Provide a description of the change below this comment. -->
Modifies the kill function to support numeric signals. I used a type check instead of `Number.isInteger` because that was the pattern used elsewhere in the code base.

Addresses: https://github.com/nodejs/node/issues/9519

addaleax commented:
Awesome! Do you think you could get a test for this together? There’s a bit on that in https://github.com/nodejs/node/blob/master/doc/guides/writing_tests.md, but I’d assume there are existing tests for the string signal names in `test/parallel` that you can add to.

And a few tiny commit message things: Could you drop the `.` at the end of the subject line, try to squeeze the subject line so that it fits into 50 characters (I know it’s hard with `child_process`… :smile:) and add a `Fixes: https://github.com/nodejs/node/issues/9519` line?

gerrard00 commented:
Will do!

addaleax commented:
Thanks, apart from the missing tests this looks good!

gerrard00 commented:
There was an existing test for child-process-kill, so I made a copy of that test and modified it to send the numeric value of SIGTERM.

I also modified the commit message as requested. I totally forgot to check that length before pushing.

cjihrig commented on lib/internal/child_process.js:
```diff
@@ -356,6 +356,8 @@ ChildProcess.prototype.kill = function(sig) {
     signal = 0;
   } else if (!sig) {
     signal = constants['SIGTERM'];
+  } else if (typeof sig === 'number') {
```
Couldn't this replace the `if (sig === 0)` check a few lines up?

gerrard00 replied:
Oh, that makes more sense. Changing now.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+'use strict';
+var common = require('../common');
```
Please use `const` instead of `var` throughout the test.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+var gotStdoutEOF = false;
+var gotStderrEOF = false;
+
+var cat = spawn(common.isWindows ? 'cmd' : 'cat');
```
Could you use `common.spawnCat()`?

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+var cat = spawn(common.isWindows ? 'cmd' : 'cat');
+
+
+cat.stdout.on('end', function() {
```
Please add `common.mustCall()` around the callback, then `gotStdoutEOF` can be completely removed.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+});
+
+cat.stderr.on('data', function(chunk) {
+  assert.ok(false);
```
Please use `common.fail()` here.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+  assert.ok(false);
+});
+
+cat.stderr.on('end', function() {
```
Please add `common.mustCall()` around the callback, then gotStderrEOF can be completely removed.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+  gotStderrEOF = true;
+});
+
+cat.on('exit', function(code, signal) {
```
Please use `common.mustCall()`, and move the assertions for `exitCode` and `termSignal` here.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+  termSignal = signal;
+});
+
+assert.equal(cat.killed, false);
```
`assert.strictEqual()` please.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+assert.equal(cat.killed, false);
+// numeric value of SIGTERM
+cat.kill(15);
+assert.equal(cat.killed, true);
```
`assert.strictEqual()` again here, please.

cjihrig commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,42 @@
+cat.kill(15);
+assert.equal(cat.killed, true);
+
+process.on('exit', function() {
```
Once the `common.mustCall()` changes are made, this can be removed completely.

gerrard00 commented:
@cjihrig Those changes are easy enough, but that will make this test very different than the existing child process kill test. Is that ok?

cjihrig commented:
The existing test is outdated, even if it's still technically correct. You can also check out the relatively new https://github.com/nodejs/node/blob/master/doc/guides/writing_tests.md.

gerrard00 commented:
@cjihrig I think I've updated the test correctly. Can you take another look when you have a moment?

Thanks for your patience!

sam-github commented:
commit messages should not have a `.` at the end of them

sam-github commented on lib/internal/child_process.js:
```diff
@@ -352,10 +352,10 @@ function onErrorNT(self, err) {
-  } else if (!sig) {
+  if (!sig) {
     signal = constants['SIGTERM'];
+  } else if (typeof sig === 'number') {
```
@cjihrig should we enforce integral, so `1.2` is not acceptable?

cjihrig replied:
@sam-github makes sense to me.

sam-github commented on lib/internal/child_process.js:
```diff
@@ -352,10 +352,10 @@ function onErrorNT(self, err) {
-  if (sig === 0) {
-    signal = 0;
-  } else if (!sig) {
+  if (!sig) {
```
did you test this? I'm pretty sure that `0` is now going to become `SIGTERM`.

targos replied:
This condition has to be after `typeof sig === 'number'` otherwise signal `0` will be incorrectly converted to `SIGTERM`

gerrard00 replied:
Ouch! That's totally wrong. I didn't test `0` after making the suggested change. Will fix.

gerrard00 replied:
Changed.

sam-github commented on test/parallel/test-child-process-kill-numeric.js:
```diff
@@ -0,0 +1,20 @@
+}));
+
+assert.strictEqual(cat.killed, false);
+// numeric value of SIGTERM
```
I think this could be `constants.os.signals.SIGTERM`, so the comment would be unnecessary.

gerrard00 replied:
Changed.

gerrard00 commented:
@sam-github The period in the commit message was removed based on feedback from @addaleax .

sam-github commented:
+1 on supporting numeric signals

sam-github commented:
I didn't look around, but if there are existing tests for kill, it would be good to add this to them, it can be hard to find tests when there is one big test file that mostly tests a feature the a couple smaller ones sprinkled about that test the incremental additions that were made over time.

^---- I didn't look around, I'm just at the github review tool right now, if its not the case that there is a good test to extend, and this new test file is the best way, pls ignore.

gerrard00 commented:
@sam-github My lunch break is over, but if I can get the `child_process` change to an acceptable state modifying `process` to support the numeric the same way should be easy.

gerrard00 commented:
@sam-github I took a cut at updating the docs and fixed a typo in that area.

cjihrig commented on doc/api/child_process.md:
```diff
@@ -789,9 +789,10 @@ within the child process to close the IPC channel as well.
 -->
 
-* `signal` {String}
+* `signal` {String|number} The signal to send, either as a string or number.
```
`String|number` -> `String|Number`

gerrard00 replied:
Fixed! That was actually a c+p from process.md, so I'll add another commit to fix that file. Or should I open a different PR for that?

cjihrig commented:
> if there are existing tests for kill, it would be good to add this to them

The test from this PR could be integrated with `test/parallel/test-child-process-kill.js` fairly easily.

gerrard00 commented:
I'll combine the new test with the existing child process test tomorrow.

sam-github commented:
Btw, I still object to adding numeric signal support to `child.kill()`, and not to `process.kill()`. Doing so creates an instantaneous bug report: "child.kill accepts numberic signals but process.kill does not". IMO, a missing feature with no bugs is better than a half-feature + a bug report. I request that either you open two PRs simultaneously, put two commits in this PR, or just do both in one commit, so the node API incrementally gets better in a single commit, instead of slightly better and slightly worse.

Also, I'm surprised child_process does its own arg checking at all, instead of calling https://github.com/nodejs/node/blob/master/lib/internal/process.js#L152-L178, so that the arg checks, Error strings, etc. would be provably identical.

I don't have time to look further, but it looks to me like `_handle.kill` could be replaced with `require('internal/process').kill`, unifying code and behaviour.

jasnell commented on lib/internal/child_process.js:
```diff
@@ -352,8 +352,8 @@ function onErrorNT(self, err) {
-  if (sig === 0) {
-    signal = 0;
+  if (typeof sig === 'number') {
+    signal = sig;
```
This should likely check that the value is not an invalid number. For instance:

```
typeof NaN === 'number'
typeof Infinity === 'number'
typeof 1.1 === 'number'
typeof -1 === 'number'
```

cjihrig replied:
`typeof sig === 'number' && sig >>> 0 === sig` should probably do the trick.

gerrard00 replied:
Awesome, you just taught me something. I've recently written similar code and used a bunch of `||` conditions.

Modified as requested.

gerrard00 commented:
@cjihrig I combined the tests as request, can you take a look?

@sam-github I'll look into using `require('internal/process').kill` and modify the code if it works. I'll also open a second PR to modify process and link to it from this PR once I've had a chance to test that.

Fishrock123 commented:
@sam-github @targos could you take a look and update your reviews accordingly?

Fishrock123 commented on doc/api/child_process.md:
```diff
@@ -1,4 +1,4 @@
-# Child Process
+#n Child Process
```
@gerrard00 You have an extra `n` here. :)

We can fix that on landing though.

Fishrock123 commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/5440/

targos commented on doc/api/child_process.md:
```diff
@@ -821,10 +821,11 @@ within the child process to close the IPC channel as well.
-The `child.kill()` methods sends a signal to the child process. If no argument
-is given, the process will be sent the `'SIGTERM'` signal. See signal(7) for
+The `child.kill()` method sends a signal to the child process. If no argument
+is given, the process will be sent the `'SIGTERM'` signal. See `signal(7)` for
```
I think we don't put backticks around man page references

sam-github replied:
is this getting fixed?

targos commented on test/parallel/test-child-process-kill.js:
```diff
@@ -1,18 +1,25 @@
+  }));
+
+  assert.equal(cat.killed, false);
+  cat.kill();
```
Shouldn't it pass `sig` here?

targos commented on test/parallel/test-child-process-kill.js:
```diff
@@ -1,18 +1,25 @@
+}
+
+runTest(constants.os.signals.SIGTERM);
+runTest('SIGTERM');
```
a third test can be added without an argument (to check the default value for signal)

sam-github commented:
My comment from https://github.com/nodejs/node/pull/9651#issuecomment-261549725 still stands:

> I still object to adding numeric signal support to child.kill(), and not to process.kill(). Doing so creates an instantaneous bug report: "child.kill accepts numeric signals but process.kill does not". IMO, a missing feature with no bugs is better than a half-feature + a bug report.

There is no reason to make these two `.kill()` APIs inconsistent. Fix them both in one PR.

Landing these kind of changes has historically been a source of node API weirdness, wherein two parts of the API subtlely diverge. There is no reason to do that here, lets just make it all better, consistently, at one time.

Fishrock123 commented:
@sam-github Ah, yes. I agree on that.

Trott commented:
@gerrard00 Are you still interested in getting this across the finish line?

Adding `stalled` for now, but if anyone (@gerrard00 or someone else) is actively working on it, please remove that label.

gerrard00 commented:
I'll close this and work on a new PR based on the feedback provided.

thefourtheye commented:
Actually #10423 covers implicitly this as well.
